src/basic/env-util.c

```diff
@@ -779,7 +779,7 @@ int serialize_environment(FILE *f, char **environment) {
                 if (!ce)
                         return -ENOMEM;
 
-                fprintf(f, "env=%s\n", *e);
+                fprintf(f, "env=%s\n", ce);
```

Comment:
I think it's a bit surprising that the result of `systemctl set-environment 'A=A\x0aB'` is different from `DefaultEnvironment=A=A\x0aB`. However, that appears to be a different issue that can be fixed or/and discussed later. I'm merging. Thank you for fixing the bug.